// @oclif/test/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.expect = exports.FancyTypes = exports.Config = exports.command = exports.test = void 0;
const fancyTest = __importStar(require("fancy-test"));
const node_path_1 = require("node:path");
const command_1 = require("./command");
const exit_1 = __importDefault(require("./exit"));
const hook_1 = __importDefault(require("./hook"));
const load_config_1 = require("./load-config");
function traverseFilePathUntil(filename, predicate) {
    let current = filename;
    while (!predicate(current)) {
        current = (0, node_path_1.dirname)(current);
    }
    return current;
}
// Update to path.dirname(url.fileURLToPath(import.meta.url)) whenever we migrate to ESM
/* eslint-disable unicorn/prefer-module */
load_config_1.loadConfig.root = traverseFilePathUntil(require.main?.path ?? module.path, (p) => !(p.includes('node_modules') || p.includes('.pnpm') || p.includes('.yarn')));
/* eslint-enable unicorn/prefer-module */
// Using a named export to import fancy causes this issue: https://github.com/oclif/test/issues/516
exports.test = fancyTest.fancy
    .register('loadConfig', load_config_1.loadConfig)
    .register('command', command_1.command)
    .register('exit', exit_1.default)
    .register('hook', hook_1.default)
    .env({ NODE_ENV: 'test' });
exports.default = exports.test;
var command_2 = require("./command");
Object.defineProperty(exports, "command", { enumerable: true, get: function () { return command_2.command; } });
var core_1 = require("@oclif/core");
Object.defineProperty(exports, "Config", { enumerable: true, get: function () { return core_1.Config; } });
var fancy_test_1 = require("fancy-test");
Object.defineProperty(exports, "FancyTypes", { enumerable: true, get: function () { return fancy_test_1.FancyTypes; } });
Object.defineProperty(exports, "expect", { enumerable: true, get: function () { return fancy_test_1.expect; } });
