// @oclif/plugin-plugins/lib/npm.js
import { ux } from '@oclif/core';
import makeDebug from 'debug';
import { readFile } from 'node:fs/promises';
import { createRequire } from 'node:module';
import { join, sep } from 'node:path';
import { fork } from './fork.js';
const debug = makeDebug('@oclif/plugin-plugins:npm');
export class NPM {
    bin;
    config;
    logLevel;
    constructor({ config, logLevel }) {
        this.config = config;
        this.logLevel = logLevel;
    }
    async exec(args = [], options) {
        const bin = await this.findNpm();
        debug('npm binary path', bin);
        args.push(`--loglevel=${this.logLevel}`, '--no-fund');
        if (this.config.npmRegistry)
            args.push(`--registry=${this.config.npmRegistry}`);
        if (options.logLevel !== 'notice' && options.logLevel !== 'silent') {
            ux.logToStderr(`${options.cwd}: ${bin} ${args.join(' ')}`);
        }
        debug(`${options.cwd}: ${bin} ${args.join(' ')}`);
        try {
            const output = await fork(bin, args, options);
            debug('npm done');
            return output;
        }
        catch (error) {
            debug('npm error', error);
            throw error;
        }
    }
    async install(args, opts) {
        const prod = opts.prod ? ['--omit', 'dev'] : [];
        return this.exec(['install', ...args, ...prod, '--no-audit'], opts);
    }
    async uninstall(args, opts) {
        return this.exec(['uninstall', ...args], opts);
    }
    async update(args, opts) {
        return this.exec(['update', ...args], opts);
    }
    async view(args, opts) {
        return this.exec(['view', ...args], { ...opts, logLevel: 'silent' });
    }
    /**
     * Get the path to the npm CLI file.
     * This will always resolve npm to the pinned version in `@oclif/plugin-plugins/package.json`.
     *
     * @returns The path to the `npm/bin/npm-cli.js` file.
     */
    async findNpm() {
        if (this.bin)
            return this.bin;
        const npmPjsonPath = createRequire(import.meta.url).resolve('npm/package.json');
        const npmPjson = JSON.parse(await readFile(npmPjsonPath, { encoding: 'utf8' }));
        const npmPath = npmPjsonPath.slice(0, Math.max(0, npmPjsonPath.lastIndexOf(sep)));
        this.bin = join(npmPath, npmPjson.bin.npm);
        return this.bin;
    }
}
