// @xata.io/cli/dist/commands/auth/login.js
import { Flags } from '@oclif/core';
import { parseProviderString } from '@xata.io/client';
import { BaseCommand } from '../../base.js';
import { buildProfile, hasProfile, setProfile } from '../../credentials.js';
class Login extends BaseCommand {
    async run() {
        var _a;
        const { flags } = await this.parseCommand();
        const profile = await this.getProfile({ ignoreEnv: true });
        const loggedIn = await hasProfile(profile.name);
        if (loggedIn) {
            const { overwrite } = await this.prompt({
                type: 'confirm',
                name: 'overwrite',
                message: `Authentication is already configured for the ${profile.name} profile, do you want to overwrite it?`
            }, flags.force);
            if (!overwrite)
                this.exit(2);
        }
        const host = parseProviderString(flags.host);
        if (!host) {
            this.error('Invalid host provider, expected either "production", "staging" or "{apiUrl},{workspacesUrl}"');
        }
        const web = flags['web-host'];
        const newProfile = buildProfile({ name: profile.name, api: flags.host, web });
        const key = (_a = flags['api-key']) !== null && _a !== void 0 ? _a : (await this.obtainKey(newProfile.web));
        await this.verifyAPIKey({ ...profile, apiKey: key, host });
        await setProfile(profile.name, { ...newProfile, apiKey: key });
        this.success('All set! you can now start using xata');
    }
}
Login.description = 'Authenticate with Xata';
Login.examples = [];
Login.flags = {
    ...BaseCommand.forceFlag('Overwrite existing credentials if they exist'),
    host: Flags.string({
        description: 'Xata API host provider'
    }),
    'web-host': Flags.string({
        description: 'Xata web host url (app.xata.io)'
    }),
    'api-key': Flags.string({
        description: 'Xata API key to use for authentication'
    })
};
Login.args = {};
export default Login;
