// @xata.io/cli/dist/commands/dbs/create.js
var _a;
import { Args, Flags } from '@oclif/core';
import { BaseCommand } from '../../base.js';
class DatabasesCreate extends BaseCommand {
    async run() {
        const { args, flags } = await this.parseCommand();
        const workspace = flags.workspace || (await this.getWorkspace());
        const result = await this.createDatabase(workspace, { overrideName: args.database, overrideRegion: flags.region });
        if (this.jsonEnabled())
            return result;
        this.success(`Database ${result.name} successfully created`);
    }
}
_a = DatabasesCreate;
DatabasesCreate.description = 'Create a database';
DatabasesCreate.examples = [];
DatabasesCreate.flags = {
    ..._a.commonFlags,
    workspace: Flags.string({
        description: 'Workspace id the database will belongs to'
    }),
    region: Flags.string({
        description: 'Region where the database will be created'
    })
};
DatabasesCreate.args = {
    database: Args.string({ description: 'The new database name', required: false })
};
DatabasesCreate.enableJsonFlag = true;
export default DatabasesCreate;
